thrust/thrust/tuple.h

```diff
@@ -178,398 +130,51 @@ get(const detail::cons<HT, TT>& t);
-  public null_type
-{
-public:
-  typedef null_type inherited;
-};
+using ::cuda::std::get;
+using ::cuda::std::make_tuple;
+using ::cuda::std::tie;
 
 /*! \endcond
  */
 
-
-/*! This version of \p make_tuple creates a new \c tuple object from a
- *  single object.
- *
- *  \param t0 The object to copy from.
- *  \return A \p tuple object with a single member which is a copy of \p t0.
- */
-template<class T0>
-__host__ __device__ inline
-  typename detail::make_tuple_mapper<T0>::type
-    make_tuple(const T0& t0);
-
-/*! This version of \p make_tuple creates a new \c tuple object from two
- *  objects.
- *
- *  \param t0 The first object to copy from.
- *  \param t1 The second object to copy from.
- *  \return A \p tuple object with two members which are copies of \p t0
- *          and \p t1.
- *
- *  \note \p make_tuple has ten variants, the rest of which are omitted here
- *        for brevity.
+/*! \} // tuple
  */
-template<class T0, class T1>
-__host__ __device__ inline
-  typename detail::make_tuple_mapper<T0, T1>::type
-    make_tuple(const T0& t0, const T1& t1);
 
-/*! This version of \p tie creates a new \c tuple whose single element is
- *  a reference which refers to this function's argument.
- *
- *  \param t0 The object to reference.
- *  \return A \p tuple object with one member which is a reference to \p t0.
+/*! \} // utility
  */
-template<typename T0>
-__host__ __device__ inline
-tuple<T0&> tie(T0& t0);
 
-/*! This version of \p tie creates a new \c tuple of references object which
- *  refers to this function's arguments.
- *
- *  \param t0 The first object to reference.
- *  \param t1 The second object to reference.
- *  \return A \p tuple object with two members which are references to \p t0
- *          and \p t1.
- *
- *  \note \p tie has ten variants, the rest of which are omitted here for
- *           brevity.
- */
-template<typename T0, typename T1>
-__host__ __device__ inline
-tuple<T0&,T1&> tie(T0& t0, T1& t1);
+THRUST_NAMESPACE_END
 
-/*! \p swap swaps the contents of two <tt>tuple</tt>s.
- *
- *  \param x The first \p tuple to swap.
- *  \param y The second \p tuple to swap.
- */
-template<
-  typename T0, typename T1, typename T2, typename T3, typename T4, typename T5, typename T6, typename T7, typename T8, typename T9,
-  typename U0, typename U1, typename U2, typename U3, typename U4, typename U5, typename U6, typename U7, typename U8, typename U9
->
-inline __host__ __device__
-void swap(tuple<T0,T1,T2,T3,T4,T5,T6,T7,T8,T9> &x,
-          tuple<U0,U1,U2,U3,U4,U5,U6,U7,U8,U9> &y);
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
 
+template<>
+struct tuple_size<tuple<THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<>> {};
```

Comment:
For all of these, since we know the number of types we do care about, couldn't we just inherit from integral_constant instead of instantiating tuple_size?

Reply:
This is essentially me being super defensive, as I fear some old compiler will hide a null_type in the initial arguments.

Would that be valid? No
Would it suprise me? No